File: Insert in a Sorted List - GFG/insert-in-a-sorted-list.js
//{ Driver Code Starts
//Initial Template for javascript


'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

class Node{
    constructor(data){
        this.data = data;
        this.next = null;
    }
}

let head = null;

function findNode(head, search_for)
{
    let current = head;
    while (current !== null)
    {
        if (current.data == search_for)
            break;
        current = current.next;
    }
    return current;
}

function printlist(head){
    let current = head; 
    let s='';
    while (current !== null)
    {
        s+=current.data+" ";
        current = current.next;
    }
    console.log(s);
}


function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let n = parseInt(readLine());
        let input_ar1 = readLine().split(' ').map(x=>parseInt(x));
        head = new Node(input_ar1[0]);
        let tail = head;
        for(let i=1;i<n;i++){
            tail.next = new Node(input_ar1[i]);
            tail = tail.next;
        }
        let data= parseInt(readLine());
        let obj = new Solution();
        let newHead = obj.sortedInsert(head, data);
        printlist(newHead);
        
    }
}


// } Driver Code Ends


//User function Template for javascript

/**
 * @param {Node} node
 * @param {number} data
 * @return {Node} node
*/


/*
class Node{
    constructor(data){
        this.data = data;
        this.next = null;
    }
}

let head;
This is method only submission.
You only need to complete the below method.
*/

class Solution {
    sortedInsert(node, data)
    {
        let new_node = new Node(data);
        
        if (node == null || node.data >= new_node.data)
        {
            // if new node is needed at the start of list
            new_node.next = node;
            return new_node;
        }
        
        let current = node;
        while (current.next != null && current.next.data < new_node.data)
            current = current.next;
            // finding a node which has lesser value than new_node
            // but its successor should have greater (or equal) value
        
        // inserting new_node after current node
        new_node.next = current.next;
        current.next = new_node;
        
        return node;
    }
}

File: Insert in a Sorted List - GFG/insert-in-a-sorted-list.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;

class Node{
    int data;
    Node next;
    
    Node(int x){
        data = x;
        next = null;
    }
    
}
class GFG{
	static void printList(Node node) 
	{ 
		while (node != null) 
		{ 
			System.out.print(node.data + " "); 
			node = node.next; 
		} 
		System.out.println(); 
	}
	
    public static void main(String args[]) throws IOException { 
        
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        
        while(t > 0){
            int n = sc.nextInt();
            
            Node head = new Node(sc.nextInt());
            Node tail = head;
            
            for(int i=0; i<n-1; i++)
            {
                tail.next = new Node(sc.nextInt());
                tail = tail.next;
            }
            
            int k = sc.nextInt();
            
            Solution g = new Solution();
            head = g.sortedInsert(head,k);
            
            printList(head); 
            t--;
        }
    } 
} 

// } Driver Code Ends


/*class Node
    {
        int data;
        Node next;
        Node(int d) {data = d; next = null; }
    }*/

// sortedInsert method should return the head of the modified linked list.
class Solution {
    Node sortedInsert(Node head, int key)
    {
        Node new_node = new Node(key);
        
        if (head == null || head.data >= new_node.data)
        {
            // if new node is needed at the start of list
            new_node.next = head;
            return new_node;
        }
        
        Node current = head;
        while (current.next != null && current.next.data < new_node.data)
            current = current.next;
            // finding a node which has lesser value than new_node
            // but its successor should have greater (or equal) value
        
        // inserting new_node after current node
        new_node.next = current.next;
        current.next = new_node;
        
        return head;
    }
}